File: src/parser.rs
use crate::ast::{Expr, Stmt};
use crate::error::ParseError;
use crate::lexer::{Token, TokenType, TokenType::*};

pub struct Parser {
    tokens: Vec<Token>,
    index: usize,
}

macro_rules! unwrap_return {
    ( $e:expr ) => {
        match $e {
            Ok(o) => o,
            Err(e) => return Err(e),
        }
    };
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Parser {
        return Parser {
            tokens,
            index: 0,
        };
    }

    pub fn parse(&mut self) -> Vec<Stmt> {
        let mut stmts: Vec<Stmt> = Vec::new();
        let mut is_error = false;

        while !self.at_end() {
            match self.statement() {
                Ok(o) => stmts.push(o),
                Err(e) => {
                    let errors: Vec<ParseError> = e.into();
                    for err in errors {
                        println!("{}", err.get_message());
                    }
                    is_error = true;
                    self.synchronize();
                }
            }
        }

        if is_error {
            return vec![];
        }

        return stmts;
    }

    fn statement(&mut self) -> Result<Stmt, Vec<ParseError>> {
        if self.match_tokens(vec![IF]) { return self.if_statement(); }
        else if self.match_tokens(vec![WHILE]) { return self.while_loop(); }
        else if self.match_tokens(vec![LOOP]) { return self.loop_stmt(); }
        else if self.match_tokens(vec![LEFT_BRACE]) { return self.block(); }
        else if self.match_tokens(vec![LET]) { return self.declaration(); }
        else if self.match_tokens(vec![FN]) { return self.function_declaration(); }
        else if self.match_tokens(vec![RETURN]) { return self.return_stmt(); }
        else if self.match_tokens(vec![CONTINUE, BREAK]) { return self.control(); }
        else if self.match_tokens(vec![CLASS]) { return self.class_declaration(); }
        else { return self.expression_statement(); }
    }

    fn class_declaration(&mut self) -> Result<Stmt, Vec<ParseError>> {
        let class_name = unwrap_return!(self.consume(VAR, "Expected class identifier"));
        unwrap_return!(self.consume(LEFT_BRACE, "Expected '{' before class body"));

        let mut methods = Vec::new();
        while !self.check(RIGHT_BRACE) && !self.at_end() {
            let stmt = unwrap_return!(self.statement());
            match stmt {
                Stmt::FnDefine { .. } => { 
                    methods.push(stmt); 
                },
                _ => { 
                    return Err(vec![ParseError::new(
                        "Expected function declaration",
                        self.previous(), 
                        self.previous().line)]) 
                }
            }
        }
        
        unwrap_return!(self.consume(RIGHT_BRACE, "expected '{' after class body"));
    
        return Ok(Stmt::ClassDefine { name: class_name, methods });
    }

    fn function_declaration(&mut self) -> Result<Stmt, Vec<ParseError>> {
        let name = unwrap_return!(self.consume(VAR, "Expected function identifier"));

        unwrap_return!(self.consume(LEFT_PAREN, "Expected ( before function arguments"));

        let mut args = Vec::new();
        if !self.check(RIGHT_PAREN) {
            args.push(unwrap_return!(self.consume(VAR, "Expected function argument")).value);
            while self.match_tokens(vec![COMMA]) {
                let arg = unwrap_return!(self.consume(VAR, "Expected function argument"));
                args.push(arg.value);
                if args.len() > 255 {
                    return Err(vec![ParseError::new(
                        "Function cannot have more than 255 arguments",
                        name.clone(),
                        name.line
                    )])
                }
            }
        }

        unwrap_return!(self.consume(RIGHT_PAREN, "Expected ) after function arguments"));

        let body = unwrap_return!(self.statement());

        return Ok(Stmt::FnDefine { name, args, body: Box::new(body), this: None });
    }

    fn return_stmt(&mut self) -> Result<Stmt, Vec<ParseError>> {
        let value = unwrap_return!(self.expression());
        unwrap_return!(self.consume(SEMI, "Expected ; after return statement"));
        return Ok(Stmt::Return { value });
    }

    fn if_statement(&mut self) -> Result<Stmt, Vec<ParseError>> {
        unwrap_return!(self.consume(LEFT_PAREN, "Expected ( after 'if'"));
        let condition = unwrap_return!(self.expression());
        unwrap_return!(self.consume(RIGHT_PAREN, "Expected ) after if condition"));

        let then = unwrap_return!(self.statement());

        let mut else_statement = None;
        if self.match_tokens(vec![ELSE]) {
            else_statement = Some(unwrap_return!(self.statement()));
        }

        return Ok(Stmt::If { 
            condition: condition, 
            then_branch: Box::new(then),
            else_branch: Box::new(else_statement)
        })
    }

    fn while_loop(&mut self) -> Result<Stmt, Vec<ParseError>> {
        unwrap_return!(self.consume(LEFT_PAREN, "Expected ( after 'if'"));
        let condition = unwrap_return!(self.expression());
        unwrap_return!(self.consume(RIGHT_PAREN, "Expected ) after if condition"));

        let body = unwrap_return!(self.statement());

        return Ok(Stmt::While { condition, body: Box::new(body) })
    }

    fn loop_stmt(&mut self) -> Result<Stmt, Vec<ParseError>> {
        let body = unwrap_return!(self.statement());

        return Ok(Stmt::Loop { body: Box::new(body) });
    }

    fn block(&mut self) -> Result<Stmt, Vec<ParseError>> {
        let mut stmts = vec![];
        let mut errors = vec![];

        while !self.check(RIGHT_BRACE) && !self.at_end() {
            let stmt = self.statement();
            match stmt {
                Ok(o) => {stmts.push(o);},
                Err(e) => {
                    for err in e {
                        errors.push(err);
                    }
                    self.synchronize();
                }
            }
        }

        unwrap_return!(self.consume(RIGHT_BRACE, "Expected '}' after block"));

        if errors.len() == 0 {
            return Ok(Stmt::Block{ statements: stmts });
        } else {
            return Err(errors);
        }
    }

    fn declaration(&mut self) -> Result<Stmt, Vec<ParseError>> {
        let name = unwrap_return!(self.consume(VAR, "Expected Identifier"));
        if self.match_tokens(vec![EQUAL]) {
            let expr = unwrap_return!(self.expression());
            unwrap_return!(self.consume(SEMI, "Expected ';' after variable declaration"));
            return Ok(Stmt::VarDefine { name, expr: Some(expr) });
        } else {
            unwrap_return!(self.consume(SEMI, "Expected ';' after variable declaration"));
            return Ok(Stmt::VarDefine { name, expr: None })
        }
    }

    fn control(&mut self) -> Result<Stmt, Vec<ParseError>> {
        let token = self.previous();
        unwrap_return!(self.consume(SEMI, "Expected semi colon after control statement"));
        return Ok(Stmt::ControlStmt { token });
    }

    fn expression_statement(&mut self) -> Result<Stmt, Vec<ParseError>> {
        let expr = unwrap_return!(self.expression());
        unwrap_return!(self.consume(SEMI, "Expected semi colon after statement"));
        return Ok(Stmt::Expression { expr });
    }

    fn expression(&mut self) -> Result<Expr, Vec<ParseError>> {
        let mut expr = unwrap_return!(self.or());

        if self.match_tokens(vec![EQUAL]) {
            match expr {
                Expr::Get { object, name } => {
                        let value = unwrap_return!(self.or());
                        expr = Expr::Set { object, field_name: name, value: Box::new(value) }
                },
                _ => {}
            }
        }

        return Ok(expr);
    }

    fn or(&mut self) -> Result<Expr, Vec<ParseError>> {
        let mut expr: Result<Expr, _> = Ok(unwrap_return!(self.and()));

        while self.match_tokens(vec![DOUBLE_LINE]) {
            let operator = self.previous();
            let right = unwrap_return!(self.and());
            expr = Ok(Expr::Logical { left: Box::new(expr.unwrap()), operator, right: Box::new(right) })
        }

        return expr;
    }

    fn and(&mut self) -> Result<Expr, Vec<ParseError>> {
        let mut expr: Result<Expr, _> = Ok(unwrap_return!(self.equality()));

        while self.match_tokens(vec![DOUBLE_AMPERSAND]) {
            let operator = self.previous();
            let right = unwrap_return!(self.equality());
            expr = Ok(Expr::Logical { left: Box::new(expr.unwrap()), operator, right: Box::new(right) })
        }

        return expr;
    }

    fn equality(&mut self) -> Result<Expr, Vec<ParseError>> {
        let mut expr = self.comparison();

        while self.match_tokens(vec![DOUBLE_EQUAL, BANG_EQUAL]) {
            let operator = self.previous();
            let right = unwrap_return!(self.comparison());
            expr = Ok(Expr::Binary {
                left: Box::new(unwrap_return!(expr)),
                operator, right: Box::new(right),
            });
        }

        return expr;
    }

    fn comparison(&mut self) -> Result<Expr, Vec<ParseError>> {
        let mut expr = self.term();

        while self.match_tokens(vec![GREATER, GREATER_EQUAL, LESS, LESS_EQUAL]) {
            let operator = self.previous();
            let right = unwrap_return!(self.term());
            expr = Ok(Expr::Binary {
                left: Box::new(unwrap_return!(expr)),
                operator,
                right: Box::new(right),
            });
        }

        return expr;
    }

    fn term(&mut self) -> Result<Expr, Vec<ParseError>> {
        let mut expr = self.factor();

        while self.match_tokens(vec![SUB, ADD]) {
            let operator = self.previous();
            let right = unwrap_return!(self.factor());
            expr = Ok(Expr::Binary {
                left: Box::new(unwrap_return!(expr)),
                operator,
                right: Box::new(right),
            });
        }

        return expr;
    }

    fn factor(&mut self) -> Result<Expr, Vec<ParseError>> {
        let mut expr = self.unary();

        while self.match_tokens(vec![MUL, DIV, MOD]) {
            let operator = self.previous();
            let right = unwrap_return!(self.unary());
            expr = Ok(Expr::Binary {
                left: Box::new(unwrap_return!(expr)),
                operator,
                right: Box::new(right),
            });
        }

        return expr;
    }

    fn unary(&mut self) -> Result<Expr, Vec<ParseError>> {
        if self.match_tokens(vec![BANG, SUB]) {
            let operator = self.previous();
            let expr = unwrap_return!(self.unary());
            return Ok(Expr::Unary {
                operator,
                right: Box::new(expr),
            });
        }

        return self.call();
    }

    fn call(&mut self) -> Result<Expr, Vec<ParseError>> {
        let mut expr = unwrap_return!(self.atom());

        loop {
            if self.match_tokens(vec![LEFT_PAREN]) {
                expr = unwrap_return!(self.finish_call(expr));
            } else if self.match_tokens(vec![DOT]) {
                let name = unwrap_return!(self.consume(VAR, "Expected identifier after '.'"));
                expr = Expr::Get { object: Box::new(expr) , name };
            } else {
                break;
            }
        }

        return Ok(expr);
    }

    fn finish_call(&mut self, expr: Expr) -> Result<Expr, Vec<ParseError>> {
        let mut args = Vec::<Expr>::new();
        if !self.check(RIGHT_PAREN) {
            args.push(unwrap_return!(self.expression()));
            while self.match_tokens(vec![COMMA]) {
                args.push(unwrap_return!(self.expression()));
            }
        }


        if args.len() > 255 {
            return Err(vec![ParseError::new(
                "Function does not support more than 255 arguments", 
                self.previous(),
                self.previous().line
            )]);
        }

        unwrap_return!(self.consume(RIGHT_PAREN, "Expected ) after function call"));

        return Ok(Expr::Call { callee: Box::new(expr), args, line: self.previous().line })
        
    }

    fn atom(&mut self) -> Result<Expr, Vec<ParseError>> {
        if self.match_tokens(vec![TRUE, FALSE, NULL, STR, INT]) {
            return Ok(Expr::Literal {
                token: self.previous(),
            });
        } else if self.match_tokens(vec![VAR]) {
            return self.identifier();
        } else if self.match_tokens(vec![SELF]) {
            return Ok(Expr::This { keyword: self.previous() });
        } else if self.match_tokens(vec![LEFT_PAREN]) {
            let expr = self.expression().unwrap();
            self.consume(RIGHT_PAREN, "Expected ) after expression").unwrap();
            return Ok(Expr::Grouping {
                expression: Box::new(expr),
            });
        }

        let token = self.peek();

        return Err(vec![ParseError::new("Expected expression", token.clone(), token.line)]);
    }

    fn identifier(&mut self) -> Result<Expr, Vec<ParseError>> {
        let name = self.previous();
        if self.match_tokens(vec![EQUAL]) {
            return Ok(Expr::VarUpdate {
                name: name,
                expr: Box::new(unwrap_return!(self.expression())),
            });
        } else {
            return Ok(Expr::Variable {
                name: name,
            });
        }
    }

    fn match_tokens(&mut self, tokens: Vec<TokenType>) -> bool {
        for token_type in tokens {
            if self.check(token_type) {
                self.advance();
                return true;
            }
        }

        return false;
    }

    fn check(&self, token_type: TokenType) -> bool {
        return self.peek().token_type == token_type;
    }

    fn at_end(&self) -> bool {
        match self.peek().token_type {
            EOF => true,
            _ => false,
        }
    }

    fn peek(&self) -> &Token {
        return self.tokens.get(self.index).unwrap();
    }

    fn previous(&self) -> Token {
        return self.tokens.get(self.index - 1).unwrap().clone();
    }

    fn advance(&mut self) -> Token {
        if !self.at_end() {
            self.index += 1;
        }
        return self.previous();
    }

    fn consume(&mut self, token_type: TokenType, message: &str) -> Result<Token, Vec<ParseError>> {
        if self.check(token_type) {
            return Ok(self.advance());
        }

        let token = self.previous();

        return Err(vec![ParseError::new(message, token.clone(), token.line)]);
    }

    fn synchronize(&mut self) {
        self.advance();
        while !self.at_end() && 
            !(self.previous().token_type == SEMI || self.previous().token_type == RIGHT_BRACE) {
            self.advance();
        }
    }
}

File: parser/parser.go
package parser

import (
	"fmt"
	. "main/errors"
	. "main/lexer"
	. "main/visitor"
	"slices"
)

var tokens []Token
var index int
var grouping int
var filename string
var text string

func Parse(token_list []Token, content string, file string) ([]Stmt, []error) {
	fmt.Print("")
	tokens = token_list
	filename = file
	text = content
	is_error := false
	stmts := []Stmt{}
	errors := []error{}
	index = 0
	grouping = 0

	for !at_end() {
		stmt, errs := declaration()
		if errs != nil {
			is_error = true
			errors = append(errors, errs...)
			synchronize()
		} else {
			stmts = append(stmts, stmt)
		}
	}

	if is_error {
		return nil, errors
	}

	return stmts, nil
}

func declaration() (Stmt, []error) {
	if match(FN) {
		return func_declaration()
	} else if match(CLASS) {
		return class_declaration()
	} else {
		return statement()
	}
}

func func_declaration() (Stmt, []error) {
	name, err := consume("Expected Identifier", peek().Location, IDENTIFIER)
	if err != nil {
		return nil, err
	}
	left_paren, err := consume("Expected '('", peek().Location, LEFT_PAREN)
	grouping++
	if err != nil {
		return nil, err
	}

	args := []string{}
	if !check(RIGHT_PAREN) {
		arg, err := consume("Expected argument name", peek().Location, IDENTIFIER)
		if err != nil {
			return nil, err
		}
		args = append(args, arg.Value.(string))
		for match(COMMA) {
			arg, err := consume("Expected argument name", peek().Location, IDENTIFIER)
			if err != nil {
				return nil, err
			}
			args = append(args, arg.Value.(string))
		}
	}

	_, err = consume("Unclosed '('", left_paren.Location, RIGHT_PAREN)
	body, err := statement()
	if err != nil {
		return nil, err
	}
	return Func{Name: name, Args: args, Body: body}, nil
}

func class_declaration() (Stmt, []error) {
	stmts := []Stmt{}
	errors := []error{}
	left := previous()

	name, err := consume("Expected identifier", peek().Location, IDENTIFIER)
	if err != nil {
		errors = append(errors, err...)
	}

	var superclass Expr = nil
	if match(COLON) {
		consume("Expected superclass name", peek().Location, IDENTIFIER)
		superclass = VarGet{Name: previous()}
	}
	_, err = consume("Expected '{'", peek().Location, LEFT_BRACE)
	if err != nil {
		errors = append(errors, err...)
	}

	match(EOL)

	for !check(RIGHT_BRACE) && !at_end() {
		stmt, err := declaration()
		if err != nil {
			errors = append(errors, err...)
			synchronize()
		}
		stmts = append(stmts, stmt)
	}

	_, err = consume("'{' was not closed", left.Location, RIGHT_BRACE)
	if err != nil {
		errors = append(errors, err...)
	}

	match(EOL)

	if len(errors) != 0 {
		return nil, errors
	} else {
		return ClassDecl{Name: name, Body: stmts, Superclass: superclass}, nil
	}
}

func statement() (Stmt, []error) {
	if match(LEFT_BRACE) {
		return block_statement()
	} else if match(TRY) {
		return try_catch_statement()
	} else if match(IF) {
		return if_statement()
	} else if match(WHILE) {
		return while_statement()
	} else if match(LOOP) {
		return loop_statement()
	} else if match(RETURN) {
		return return_statement()
	} else if match(BREAK) {
		return break_statement()
	} else if match(CONTINUE) {
		return continue_statement()
	} else if match(IMPORT) {
		return import_statement()
	} else {
		return expression_statement()
	}
}

func block_statement() (Stmt, []error) {
	stmts := []Stmt{}
	errors := []error{}
	left := previous()

	match(EOL)

	for !check(RIGHT_BRACE) && !at_end() {
		stmt, err := statement()
		if err != nil {
			errors = append(errors, err...)
			synchronize()
		}
		stmts = append(stmts, stmt)
	}

	_, err := consume("'{' was not closed", left.Location, RIGHT_BRACE)
	if err != nil {
		errors = append(errors, err...)
	}

	match(EOL)

	if len(errors) != 0 {
		return nil, errors
	} else {
		return Block{Body: stmts}, nil
	}
}

func try_catch_statement() (Stmt, []error) {
	try := previous()
	try_body, err := statement()
	if err != nil {
		return nil, err
	}
	_, err = consume("Incomplete Try Catch statement", try.Location, CATCH)
	if err != nil {
		return nil, err
	}

	identifier, _ := consume("Expected identifier", peek().Location, IDENTIFIER)

	catch_body, err := statement()
	if err != nil {
		return nil, err
	}

	return TryCatch{TryBody: try_body, CatchBody: catch_body,
		ErrName: identifier}, nil
}

func if_statement() (Stmt, []error) {
	expr, err := expression()
	if err != nil {
		return nil, err
	}
	body, err := statement()
	if err != nil {
		return nil, err
	}

	var else_body Stmt = nil
	if match(ELSE) {
		else_body, err = statement()
		if err != nil {
			return nil, err
		}
	}

	return If{IfBody: body, ElseBody: else_body, Condition: expr}, nil
}

func while_statement() (Stmt, []error) {
	expr, err := expression()
	if err != nil {
		return nil, err
	}

	body, err := statement()
	if err != nil {
		return nil, err
	}

	return While{Body: body, Condition: expr}, nil
}

func loop_statement() (Stmt, []error) {
	body, err := statement()
	if err != nil {
		return nil, err
	}

	return Loop{Body: body}, nil
}

func return_statement() (Stmt, []error) {
	if tokens[index-1].Type == EOL {
		return Return{Expr: Literal{Value: nil}, Line: previous().Line}, nil
	}
	expr, err := expression()
	if err != nil {
		return nil, err
	}

	_, err = consume("Invalid Syntax", peek().Location, EOL)

	return Return{Expr: expr, Line: previous().Line}, nil
}

func break_statement() (Stmt, []error) {
	_, err := consume("Invalid Syntax", peek().Location, EOL)
	if err != nil {
		return nil, err
	}
	return Break{Line: previous().Line}, nil
}

func continue_statement() (Stmt, []error) {
	_, err := consume("Invalid Syntax", peek().Location, EOL)
	if err != nil {
		return nil, err
	}
	return Continue{Line: previous().Line}, nil
}

func import_statement() (Stmt, []error) {
	path, err := consume("Expected import path", peek().Location, STR)
	if err != nil {
		return nil, err
	}
	_, err = consume("Invalid syntax", previous().Location, EOL)
	if err != nil {
		return nil, err
	}
	return Import{Path: path}, nil
}

func expression_statement() (Stmt, []error) {
	expr, err := expression()
	if err != nil {
		return nil, err
	}
	_, err = consume("Invalid Syntax", peek().Location, EOL)
	if err != nil {
		return nil, err
	}

	return Expression{Expr: expr}, nil
}

func expression() (Expr, []error) {
	return assignment()
}

func assignment() (Expr, []error) {
	local := check(LOCAL)
	if local {
		advance()
	}
	expr, err := compound_assignment()
	if err != nil {
		return nil, err
	}
	if match(EQUAL) {
		equals := previous()
		val, err := assignment()
		if err != nil {
			return nil, err
		}
		switch expr := expr.(type) {
		case VarGet:
			return VarDef{Name: expr.Name, Expr: val, Local: local}, nil
		case Get:
			return Set{Expr: expr.Expr, Name: expr.Name, Val: val}, nil
		case IterGet:
			return IterSet{Expr: expr.Expr, Index: expr.Index, 
				Line: expr.Line, Val: val}, nil
		default:
			return nil, []error{SyntaxError("Invalid assignment target", equals.Line,
				equals.Location, text, filename)}
		}
	} else if local {
		return nil, []error{SyntaxError("Expected variable name", peek().Line,
			peek().Location, text, filename)}
	}

	return expr, nil
}

func compound_assignment() (Expr, []error) {
	expr, err := tuple()
	if err != nil {
		return nil, err
	}

	if match(ADD_EQUAL, SUB_EQUAL, MUL_EQUAL, DIV_EQUAL, MOD_EQUAL) {
		equals := previous()
		equals.Type -= 1
		val, err := compound_assignment()
		if err != nil {
			return nil, err
		}
		switch expr := expr.(type) {
		case VarGet:
			return VarDef{
				Name: expr.Name,
				Expr: Binary{
					Left:     expr,
					Operator: equals,
					Right:    val,
				},
			}, nil
		case Get:
			return Set{
				Name: expr.Name,
				Expr: expr.Expr,
				Val: Binary{
					Left:     expr,
					Operator: equals,
					Right:    val,
				},
			}, nil
		case IterGet:
			return IterSet{
				Expr: expr.Expr,
				Index: expr.Index,
				Line: expr.Line,
				Val: Binary{
					Left: expr,
					Operator: equals,
					Right: val,
				},
			}, nil
		default:
			return nil, []error{SyntaxError("Invalid assignment target", equals.Line,
				equals.Location, text, filename)}
		}
	}

	return expr, nil
}

func tuple() (Expr, []error) {
	expr, err := logic_or()
	if err != nil {
		return nil, err
	}

	if check(COMMA) {
		exprs := []Expr{expr}
		for match(COMMA) {
			val, err := logic_or()
			if err != nil {
				return nil, err
			}
			exprs = append(exprs, val)
		}

		expr = Tuple{Vals: exprs}
	}

	return expr, nil
}

func logic_or() (Expr, []error) {
	expr, err := logic_and()
	if err != nil {
		return nil, err
	}

	for match(OR) {
		operator := previous()
		right, err := logic_and()
		if err != nil {
			return nil, err
		}
		expr = Logical{Left: expr, Operator: operator, Right: right}
	}

	return expr, nil
}

func logic_and() (Expr, []error) {
	expr, err := equality()
	if err != nil {
		return nil, err
	}

	for match(AND) {
		operator := previous()
		right, err := equality()
		if err != nil {
			return nil, err
		}
		expr = Logical{Left: expr, Operator: operator, Right: right}
	}

	return expr, nil
}

func equality() (Expr, []error) {
	var expr, err = comparison()
	if err != nil {
		return nil, err
	}

	for match(EQUAL_EQUAL, BANG_EQUAL) {
		var operator = previous()
		var right, err = comparison()
		if err != nil {
			return nil, err
		}
		expr = Truth{Left: expr, Operator: operator, Right: right}
	}

	return expr, nil
}

func comparison() (Expr, []error) {
	var expr, err = term()
	if err != nil {
		return nil, err
	}

	for match(GREATER, GREATER_EQUAL, LESS, LESS_EQUAL) {
		var operator = previous()
		var right, err = term()
		if err != nil {
			return nil, err
		}
		expr = Truth{Left: expr, Operator: operator, Right: right}
	}

	return expr, nil
}

func term() (Expr, []error) {
	var expr, err = factor()
	if err != nil {
		return nil, err
	}

	for match(ADD, SUB) {
		var operator = previous()
		var right, err = factor()
		if err != nil {
			return nil, err
		}
		expr = Binary{Left: expr, Operator: operator, Right: right}
	}

	return expr, nil
}

func factor() (Expr, []error) {
	var expr, err = power()
	if err != nil {
		return nil, err
	}

	for match(MUL, DIV, MOD) {
		var operator = previous()
		var right, err = power()
		if err != nil {
			return nil, err
		}
		expr = Binary{Left: expr, Operator: operator, Right: right}
	}

	return expr, nil
}

func power() (Expr, []error) {
	expr, err := unary()
	if err != nil {
		return nil, err
	}

	for match(POWER) {
		operator := previous()
		right, err := unary()
		if err != nil {
			return nil, err
		}
		expr = Binary{Left: expr, Operator: operator, Right: right}
	}

	return expr, nil
}

func unary() (Expr, []error) {
	if match(BANG, SUB, ADD) {
		var operator = previous()
		var expr, err = unary()
		if err != nil {
			return nil, err
		}
		return Unary{Operator: operator, Expr: expr}, nil
	}

	return list()
}

func list() (Expr, []error) {
	expr, err := call()
	if err != nil {
		return nil, err
	}

	for true {
		if match(LEFT_SQUARE) {
			left_square := previous()
			grouping++
			val, err := logic_or()
			if err != nil {
				return nil, err
			}
			consume("Unclosed [", left_square.Location, RIGHT_SQUARE)
			expr = IterGet{Expr: expr, Index: val, Line: previous().Line}
		} else {
			break
		}
	}

	return expr, nil
}

func call() (Expr, []error) {
	expr, err := atom()
	if err != nil {
		return nil, err
	}

	for true {
		if match(LEFT_PAREN) {
			left_paren := previous()
			grouping++
			args := []Expr{}
			if !check(RIGHT_PAREN) {
				arg, err := logic_or()
				if err != nil {
					return nil, err
				}
				args = append(args, arg)
				for match(COMMA) {
					arg, err := logic_or()
					if err != nil {
						return nil, err
					}
					args = append(args, arg)
				}
			}
			var right_paren Token
			if check(RIGHT_PAREN) {
				index++
				right_paren = previous()
			} else {
				return nil, []error{SyntaxError("Unclosed '('",
					left_paren.Line, left_paren.Location, text, filename)}
			}
			if err != nil {
				return nil, err
			}
			grouping--
			expr = Call{Callee: expr, Args: args, Paren: right_paren}
		} else if match(DOT) {
			name, err := consume("Expected property name",
				peek().Location, IDENTIFIER)
			if err != nil {
				return nil, err
			}
			expr = Get{Expr: expr, Name: name}
		} else {
			break
		}
	}

	return expr, nil
}

func atom() (Expr, []error) {
	if match(FALSE) {
		return Literal{Value: false}, nil
	} else if match(TRUE) {
		return Literal{Value: true}, nil
	} else if match(NULL) {
		return Literal{Value: nil}, nil
	} else if match(LEFT_SQUARE) {
		left_square := previous()
		exprs := []Expr{}
		if !check(RIGHT_SQUARE) {
			grouping++
			expr, err := logic_or()
			if err != nil {
				return nil, err
			}
			exprs = append(exprs, expr)
			if check(COMMA) {
				for match(COMMA) {
					val, err := logic_or()
					if err != nil {
						return nil, err
					}
					exprs = append(exprs, val)
				}

			}
		}
		_, err := consume("Unclosed [", left_square.Location, RIGHT_SQUARE)
		if err != nil {
			return nil, err
		}

		grouping--
		return ListDef{Vals: exprs}, nil
	} else if match(NUM, STR) {
		return Literal{Value: previous().Value}, nil
	} else if match(IDENTIFIER) {
		return VarGet{Name: previous()}, nil
	} else if match(SELF) {
		return Self{Line: previous().Line}, nil
	} else if match(LEFT_PAREN) {
		grouping++
		var left_paren = previous()
		var expr, err = expression()
		if err != nil {
			return nil, err
		}
		consume("'(' was not closed", left_paren.Location, RIGHT_PAREN)
		grouping--
		return Grouping{Expr: expr}, nil
	}

	return nil, []error{SyntaxError("Expected Expression", peek().Line,
		peek().Location, text, filename)}
}

func match(types ...TokenType) bool {
	if slices.Contains(types, peek().Type) {
		advance()
		return true
	}
	return false
}

func advance() Token {
	if !at_end() {
		index++
	}
	if (grouping > 0 && peek().Type == EOL ||
		previous().Type == EOL && peek().Type == EOL) &&
		!at_end() {
		return advance()
	}
	return previous()
}

func at_end() bool {
	temp := index
	for tokens[temp].Type != EOF {
		temp++
		if tokens[temp-1].Type != EOL {
			return false
		}
	}
	return true
}

func peek() Token {
	return tokens[index]
}

func previous() Token {
	if grouping > 0 {
		temp := -1
		for (temp+index) > 0 && tokens[index+temp].Type == EOL {
			temp -= 1
		}
		return tokens[index+temp]
	} else {
		return tokens[index-1]
	}
}

func consume(err string, location int, token_type TokenType) (Token, []error) {
	if match(token_type) {
		return previous(), nil
	}

	// Kinda a bandaid solution. Removing semi colons is a pain in the ass
	if token_type == EOL && tokens[index-1].Type == EOL {
		return previous(), nil
	}

	// I can't return nil since this function returns a struct,
	// but the results of this function are discared if an error is thrown
	return previous(), []error{SyntaxError(err, previous().Line,
		location, text, filename)}
}

func check(token_type TokenType) bool {
	return peek().Type == token_type
}

func synchronize() {
	grouping = 0
	advance()

	for !at_end() {
		if previous().Type == EOL {
			return
		}

		switch peek().Type {
		case CLASS, FN, FOR, WHILE, IF, RETURN:
			return
		}

		advance()
	}
}
